sopel/coretasks.py

```diff
@@ -55,10 +60,105 @@
+        )
+
+    return plugin.CapabilityNegotiation.DONE
+
+
+def _sasl_capability(
+    cap_req: Tuple[str, ...], bot: SopelWrapper, acknowledged: bool,
+) -> plugin.CapabilityNegotiation:
+    # Manage CAP REQ :sasl
+    auth_method = bot.settings.core.auth_method
+    server_auth_method = bot.settings.core.server_auth_method
+    is_required = 'sasl' in (auth_method, server_auth_method)
+
+    if not is_required:
+        # not required: we are fine, available or not
+        return plugin.CapabilityNegotiation.DONE
+    elif not acknowledged:
+        # required but not available: error, we must stop here
+        LOGGER.error(
+            'SASL capability is not enabled; '
+            'cannot authenticate with SASL.',
+        )
+        return plugin.CapabilityNegotiation.ERROR
+
+    # Check SASL configuration (password is required)
+    password, mech = _get_sasl_pass_and_mech(bot)
+    if not password:
+        raise config.ConfigurationError(
+            'SASL authentication required but no password available; '
+            'please check your configuration file.',
+        )
+
+    cap_info = bot.capabilities.get_capability_info('sasl')
+    cap_params = cap_info.params
+
+    available_mechs = cap_params.split(',') if cap_params else []
+
+    if available_mechs and mech not in available_mechs:
+        # Raise an error if configured to use an unsupported SASL mechanism,
+        # but only if the server actually advertised supported mechanisms,
+        # i.e. this network supports SASL 3.2
 
+        # SASL 3.1 failure is handled (when possible)
+        # by the sasl_mechs() function
 
-def setup(bot):
+        # See https://github.com/sopel-irc/sopel/issues/1780 for background
+        raise config.ConfigurationError(
+            'SASL mechanism "{mech}" is not advertised by this server; '
+            'available mechanisms are: {available}.'.format(
+                mech=mech,
+                available=', '.join(available_mechs),
+            )
+        )
+
+    bot.write(('AUTHENTICATE', mech))
+
+    # If we want to do SASL, we have to wait before we can send CAP END. So if
+    # we are, wait on 903 (SASL successful) to send it.
+    return plugin.CapabilityNegotiation.CONTINUE
+
+
+CAP_ECHO_MESSAGE = plugin.capability('echo-message')
+CAP_MULTI_PREFIX = plugin.capability('multi-prefix')
+CAP_AWAY_NOTIFY = plugin.capability('away-notify')
+CAP_CHGHOST = plugin.capability('chghost')
+CAP_CAP_NOTIFY = plugin.capability('cap-notify')
+CAP_SERVER_TIME = plugin.capability('server-time')
+CAP_USERHOST_IN_NAMES = plugin.capability('userhost-in-names')
+CAP_MESSAGE_TAGS = plugin.capability('message-tags')
+CAP_ACCOUNT_NOTIFY = plugin.capability(
+    'account-notify', handler=_auth_capability)
+CAP_EXTENDED_JOIN = plugin.capability(
+    'extended-join', handler=_auth_capability)
+CAP_ACCOUNT_TAG = plugin.capability(
+    'account-tag', handler=_auth_capability)
+CAP_SASL = plugin.capability('sasl', handler=_sasl_capability)
```

Comment:
Clearly I sat on this for a while. 😁 Didn't want to rush into saying any more on the topic, as it's clearly a _very big_ part of how this will look from a plugin author's perspective.

And in that time, I landed on _what_ bugs me about how it looks now: The top level of the plugin's module gets cluttered.

So the second way you listed to expose properties (an object/class or `Protocol`) sounds much better to me. Heck, if we land on something that can be adapted to a "version 2" of the `StaticSection` idiom for settings, even more better.

I think a main sticking point for us would be introducing another reserved name like `setup`, `shutdown`, and `configure`, right? Even if it's a reserved class name, not a function name. Leaving aside that it's _probably_ fine to add a reserved name in a major version, let me run with the alternatives for a bit. Hope you don't mind a bit of brainstorming-out-loud.

### Option 1: Class registered at `setup()`

- In brief: Like `config.types.StaticSection`, but for capability requests.
- Usage model: Plugin defines a `PluginNameCAPs` class (or whatever the convention ends up being), then registers it in `setup()` like how `PluginNameSection` classes get registered for settings now. Class members could look like the top-level module attributes do now in this revision of `coretasks`, i.e. each member/attr is a `plugin.capability`.
- **Pros:**
  - Familiar idiom, to anyone who's written even a moderately complex Sopel plugin
  - Creates only one name at the module top level
  - Can be named anything the author wants
  - Multiple instances possible, if someone wants to go totally nuts
    - For example, register the plugin's settings first and then conditionally add CAP requests based on the loaded settings
- **Cons:**
  - I (at least) want to move _away_ from that exact model of how config sections are registered in `setup()`
    - Not sure yet if the CAP version of this approach would require passing the plugin's name again, but that's something I/we don't like about `bot.config.add_section()`
  - I had included "naming attrs is hard" (but more verbose/technical) here, but then realized that the attr names literally don't matter as long as the attrs are self-describing, i.e. `plugin.capability` objects

### Option 2: Abstract class to extend in plugin

- In brief: Also like `config.types.StaticSection`, but no need to register in `setup()`.
- Usage model: Plugin defines `PluginNameCAPs(plugin.Capabilities)` (whatever the names are) and the loader/plugin-handler auto-collects all subclasses of the core-defined type. Members of this class are the same as Option 1.
- **Pros:**
  - Also creates only one name at the module top level
  - Also can be named anything
  - Also multiple instances possible
    - Less useful here because the `bot.settings` won't be available for conditional enabling/disabling of individual requests
  - Could be adapted to plugin config sections pretty easily (`config.types.AutoSection` in Sopel 9.0, anyone?)
    - The plugin loader/handler already knows the plugin's name, so this can eliminate needing to repeat it when calling `bot.config.add_section()`
- **Cons:**
  - …I'm sure they exist, but this fixes the cons I listed for Option 1 and I couldn't think of any specific to Option 2.

### Conclusion… for now?

I didn't write up an option for a reserved name because it would be basically the same as Option 2 but with a predefined name the plugin author has to use instead of extending a specific class/interface.

Now I wait and see what (if anything) survives The Exirel Gauntlet, i.e. how many of these ideas are Bad™ based on your much greater past experience in software engineering. 🙃

Reply:
Since I spent times away from writing code for a while, I had time to think about this. I still want the same thing, i.e. declaration of plugin's features at import time, not by running `setup`. And we still have the same constraint on CAP: requesting cap1 and cap2 separately is *not* the same thing as requesting them together.

This means that using a single class would be quite annoying: you would have to do something like that:

```python
class PluginCapRequests():
    CAP_REQUESTS = (
        ('cap1',),
        ('cap2', 'cap3'),
    )

    def cap_is_ack(self, cap_req):
        # check what is in cap_req so you know which CAP_REQUESTS is ack
```

All it does is moving complexity into the handling custom code. Another option would be to do something like:

```python
class PluginCapRequests():
    @plugin.capability('cap1')
    def cap1_is_ack(self, ...):
        ...

    @plugin.capability('cap2', 'cap3')
    def cap2_3_is_ack(self, ...):
        ...
```

Which is exactly like having a bunch a module-level attribute, but now you *must* have a CAP handler to decorate.

Last but not least: how many plugins are going to need that? Like, I get that you don't like the view, but there is only one plugin that need that: `coretasks`. This clutter is ours only, for a feature that is so specific, so advanced, I don't see how the average plugin author will get confused with it.

So, even if I considered changing to something slightly different for a moment, I would rather stick with what I suggest in this PR for now.